e2e/testcases/oci_signature_verification_test.go

```diff
@@ -0,0 +1,265 @@
+	}
+
+	nt.T.Log("Check that verification is up-to-date with floating tag")
+	nt.T.Logf("Checking no signature error exists for second image with digest %s", image1.Digest)
+	nt.Must(
+		nt.Watcher.WatchObject(kinds.RootSyncV1Beta1(), rootSyncID.Name, configsync.ControllerNamespace,
+			testwatcher.WatchPredicates(
+				testpredicates.RootSyncHasSourceError(status.SourceErrorCode, "no signatures found"),
+				testpredicates.RootSyncHasSourceError(status.SourceErrorCode, image1.Digest))))
+
+	if err = signImage(nt, image1); err != nil {
+		nt.T.Fatalf("Failed to sign second test image %s", err)
+	}
+	nt.T.Log("Update source expectation for watch")
+	_ = nt.RootSyncObjectOCI(rootSyncKey.Name, image.OCIImageID().WithoutDigest(), "", image1.Digest)
+	nt.Must(nt.WatchForAllSyncs())
+	nt.T.Log("Check RootSync is synced to second image and annotations are updated.")
+	err = nt.Watcher.WatchObject(kinds.RootSyncV1Beta1(), rootSyncID.Name, rootSyncID.Namespace,
+		testwatcher.WatchPredicates(
+			checkRootSyncPreSyncAnnotations(image1),
+		))
+	if err != nil {
+		nt.T.Fatalf("Source annotation not updated for RootSync %v", err)
+	}
+
+	nt.T.Log("Set the RootSync to sync from Git")
+	nt.SyncSources[rootSyncID] = gitSource
+	rootSyncGit := nomostest.RootSyncObjectV1Beta1FromRootRepo(nt, rootSyncID.Name)
+	nt.Must(nt.KubeClient.Apply(rootSyncGit))
+	nt.Must(nt.WatchForAllSyncs())
+
+	err = nt.Watcher.WatchObject(kinds.RootSyncV1Beta1(), rootSyncID.Name, rootSyncID.Namespace,
+		testwatcher.WatchPredicates(
+			testpredicates.MissingAnnotation(metadata.ImageToSyncAnnotationKey),
+		))
+	if err != nil {
+		nt.T.Fatalf("Source annotations still exist when RootSync is syncing from Git %v", err)
+	}
+}
+
+func checkRootSyncPreSyncAnnotations(image *registryproviders.OCIImage) testpredicates.Predicate {
+	return func(o client.Object) error {
+		if o == nil {
+			return testpredicates.ErrObjectNotFound
+		}
+		rs, ok := o.(*v1beta1.RootSync)
+		if !ok {
+			return testpredicates.WrongTypeErr(o, &v1beta1.RootSync{})
+		}
+		syncingCondition := rootsync.GetCondition(rs.Status.Conditions, v1beta1.RootSyncSyncing)
+		syncingCommit := syncingCondition.Commit
+		syncingImage := rs.Spec.Oci.Image
+		imageToSyncAnnotation, ok := o.GetAnnotations()[metadata.ImageToSyncAnnotationKey]
+		if !ok {
+			return fmt.Errorf("object %q does not have annotation %q", o.GetName(), metadata.ImageToSyncAnnotationKey)
+		}
+		remoteAddr := image.OCIImageID().Address()
+		if imageToSyncAnnotation != fmt.Sprintf("%s@sha256:%s", syncingImage, syncingCommit) || imageToSyncAnnotation != remoteAddr {
+			return fmt.Errorf("expecting to have OCI image %s, but got image-to-sync annotation %s, syncing commit %s, syncing image %s",
+				remoteAddr,
+				imageToSyncAnnotation,
+				syncingCommit,
+				syncingImage,
+			)
+		}
+		return nil
+	}
+}
+
+func checkRepoSyncPreSyncAnnotations(image *registryproviders.OCIImage) testpredicates.Predicate {
```

Comment:
Do we need these two functions?

We can use `testpredicates.HasAnnotation(metadata.ImageToSyncAnnotationKey, image1.OCIImageID().Address())` to replace them.

Reply:
Done